XiEditor/CoreConnection.swift

```diff
@@ -150,29 +146,82 @@ class CoreConnection {
                     callback = self.pending.removeValue(forKey: index)
                 }
                 callback?(result)
-            } else { // is request
-                DispatchQueue.main.async {
-                    let result = self.callback(json as AnyObject)
-                    let resp = ["id": index, "result": result] as [String : Any?]
-                    self.sendJson(resp as Any)
-                }
-            }
-            // is notification
-        } else {
-            // updates and style defs get their own codepath, staying on this thread;
-            // the main thread may be blocked waiting for this update
-            if let method = obj["method"] as? String {
-                if method == "update" || method == "def_style", let params = obj["params"] as? [String: AnyObject] {
-                    self.updateCallback(method, params)
-                } else {
-                    // other notifications go on the main thread
-                    DispatchQueue.main.async {
-                        let _ = self.callback(json as AnyObject)
-                    }
-                }
             } else {
-                print("malformed json-rpc notification \(json)")
+                self.handleRequest(json: obj)
             }
+        } else {
+            self.handleNotification(json: obj)
+        }
+    }
+
+    func handleRequest(json: [String: AnyObject]) {
+        // there are currently no core -> client requests in the protocol
+        print("Unexpected RPC Request: \(json)")
+    }
+
+    func handleNotification(json: [String: AnyObject]) {
```

Comment:
I think at some point I'd like to see the low-level json-rpc stuff split out from the specific methods handled, but for now that's very much yagni, as there's only one interface the client needs to implement directly.

Reply:
I.e. you mean that the parsing of _specific_ methods should live outside of core connection? I agree that would make sense.

Reply:
Yes, that's what I meant.